### PR TITLE
fix: change the fetch query so that the filter is passed

### LINKED ISSUE #3614: [Grid] In memory Grid with SelectAllCheckboxVisibility.VISIBLE ignores filter on selectAll
### Description

I have a Grid with a ListDataProvider.
I use ListDataView.setFilter() to set a filter.
The SelectAllCheckBox is visible by default.
The filter filters out some items of the ListDataView.
But a click on the selectAll checbox selects all items in the ListDataView. getSelectedItems() returns all items.

### Expected outcome

A click on the SelectAll CheckBox should not select the filtered out items.

### Minimal reproducible example

````java
private Grid<Data> createGrid() {
        final List<Data> data = new ArrayList<>();
        for(int i=1; i<=5; i++) data.add(new Data(i,"Name"+i));

        final Grid<Data> grid = new Grid<>();
        grid.addColumn(Data::getId).setHeader("Id");
        grid.addColumn(Data::getName).setHeader("Name");
        grid.setSelectionMode(Grid.SelectionMode.MULTI);
        ListDataView<Data, ?> dataView = grid.setItems(data);
        dataView.setFilter(d -> d.getId() < 3);
        grid.addSelectionListener(e -> System.out.println("Selected count: " + e.getAllSelectedItems().size()));
        return grid;
    }
````

### Steps to reproduce

1. Add Grid above to view
2. Click on select all CheckBox
3. Console shows 'Selected count: 5' instead of 'Selected count: 2'

### Environment

Vaadin version(s): 23.1.6
OS: Windows

### Browsers

_No response_

### PR DESCRIPTION
**Pull request by ugur-vaadin**
## Description

For a multi-select Grid, when a filter was introduced using the ListDataView, using the select all check box from the client doesn't take the filter into account. This PR changes the query so that the query contains the filter info that is already present in the DataCommunicator.

Fixes #3614 

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java [hunk 1]
```diff
@@ -404,6 +404,29 @@ public void updateSelection_updatesCheckboxStates() {
                 (boolean) columnElement.getPropertyRaw("indeterminate"));
     }
 
+    @Test
+    public void setFilterUsingDataView_clientSelectAll_selectionEventContainsFilteredValues() {
+        grid.setSelectionMode(SelectionMode.MULTI);
+        List<String> items = List.of("foo", "bar");
+        ListDataProvider<String> dataProvider = DataProvider
+                .ofCollection(items);
+        ListDataView<String, ?> dataView = grid.setItems(dataProvider);
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java:413]:
```suggestion
        ListDataView<String, ?> dataView = grid.setItems(items);
```

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 2]
```diff
@@ -375,7 +375,8 @@ void clientSelectAll() {
             allItemsStream = fetchAllHierarchical(
                     (HierarchicalDataProvider<T, ?>) dataProvider);
         } else {
-            allItemsStream = dataProvider.fetch(new Query<>());
+            allItemsStream = dataProvider.fetch(getGrid().getDataCommunicator()
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:378]: My initial thought is that this should also be applied in `AbstractGridMultiSelectionModel.selectAll` / when using server-side selection.

Comment (ugur-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:380]: This is true. Addressed in the new commit.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java (hunk 1 above)
Comment (ugur-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java]: Addressed in the new commit.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 3]
```diff
@@ -209,8 +209,10 @@ public void deselect(T item) {
     @Override
     public void selectAll() {
         updateSelection(
-                getGrid().getDataCommunicator().getDataProvider()
-                        .fetch(new Query<>()).collect(Collectors.toSet()),
+                (Set<T>) getGrid().getDataCommunicator().getDataProvider()
+                        .fetch(getGrid().getDataCommunicator().buildQuery(0,
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:213]: Let's add a test for this as well.

Comment (ugur-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:213]: Separated the test cases in the new commit.

Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:213]: Sorry, I missed that they were combined, but extracting a separate test is an improvement 👍.

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**